### PR TITLE
crimson/osd/shard_services: fix dangling reference caused by rvalue reference of ShardSercies::dispatch_context()

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/68662
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/shard_services.cc [hunk 1]
```diff
@@ -802,8 +802,9 @@ seastar::future<> ShardServices::dispatch_context_messages(
 
 seastar::future<> ShardServices::dispatch_context(
   crimson::os::CollectionRef col,
-  PeeringCtx &&ctx)
+  PeeringCtx &&pctx)
 {
+  auto ctx = std::move(pctx);
```
Comment (Matan-B) [inline src/crimson/osd/shard_services.cc:807]: I'm not sure I'm understanding this change. Can you please explain?
Looking at `dispatch_context_transaction` (from the backtrace), it seems that we should be passing rval (`PeeringCtx &&ctx`) as parameter and not as a reference, right? Could that explain the tracker?

Comment (xxhdx1985126) [reply src/crimson/osd/shard_services.cc]: Yeah, but not enough, I've just updated this PR.
Passing rval as parameter can solve this issue if `ShardServices::dispatch_context()` is also a coroutine. But in this case, it's not, so I used `seastar::do_with()` to hold `PeeringCtx` in the latest update. Please take a look.

File: src/crimson/osd/shard_services.cc [hunk 2]
```diff
@@ -802,15 +802,19 @@ seastar::future<> ShardServices::dispatch_context_messages(
...
-  return seastar::when_all_succeed(
-    dispatch_context_messages(
-      BufferedRecoveryMessages{ctx}),
-    col ? dispatch_context_transaction(col, ctx) : seastar::now()
-  ).then_unpack([] {
-    return seastar::now();
+  PeeringCtx &&pctx)
+{
+  return seastar::do_with(
+    std::move(pctx),
```
Comment (Matan-B) [inline src/crimson/osd/shard_services.cc:808]: lgtm! Got it now, the passed `ctx` instance is local (to dispatch_context) and should be kept alive as we use it inside the coroutine.

File: src/crimson/osd/shard_services.cc [hunk 3]
```diff
@@ -802,15 +802,19 @@ seastar::future<> ShardServices::dispatch_context_messages(
...
-  PeeringCtx &&ctx)
-{
-  ceph_assert(col || ctx.transaction.empty());
-  return seastar::when_all_succeed(
-    dispatch_context_messages(
-      BufferedRecoveryMessages{ctx}),
-    col ? dispatch_context_transaction(col, ctx) : seastar::now()
-  ).then_unpack([] {
-    return seastar::now();
+  PeeringCtx &&pctx)
```
Comment (Matan-B) [inline src/crimson/osd/shard_services.cc:805]: nit: rename header declaration as well

**Review by Matan-B — APPROVED** [review]

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-nov-3/

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-11-03_10:45:16-crimson-rados-wip-matanb-crimson-testing-nov-3-distro-crimson-smithi/

7978413, 7978439 - https://tracker.ceph.com/issues/67791
Seastore:
7978417 - https://tracker.ceph.com/issues/67823
7978430 - https://tracker.ceph.com/issues/68809
Infra:
7978415 - Error reimaging machines

**Pull request merged**